### ISSUE TITLE
feat: Allow integration of Angular custom form controls with ion-item validation behavior

### ISSUE THREAD

**Issue by gmaughan** [ionic-team/ionic-framework#21903]
# Feature Request

**Ionic version:**

[x] 5.3.1

**Describe the Feature Request**

Ionic/Angular has a nice UI feature that presents validation status to users in the form of CSS classes that map to the ng form control classes. eg,` ng-valid/ng-invalid` maps to `ion-valid/ion-invalid` etc. And these classes somehow get pushed to the containing `ion-item` of the control which does the actual presentation.

We have several custom form control components and we would like to have them behave in the same manner. But there doesn't seem to be an easy way to integrate with this functionality.

Please provide a mechanism to integrate custom form controls with the ion-item valid/invalid presentation behavior.

We currently cannot implement this behavior and our forms do not function well when custom form controls are used within `ion-item` lists.

**Describe Preferred Solution**

An abstract version of Ionic's [`ValueAccessor`](https://github.com/ionic-team/ionic-framework/blob/master/angular/src/directives/control-value-accessors/value-accessor.ts) class that components could extend would be nice. This seems to be where the magic happens.

**Describe Alternatives**

Alternatively, a number of public hooks that could be called to patch the Angular functions and set the Ionic classes would also work.

**Comment by gmaughan** [comment]
Here is an abstract class that almost works in that it sets the classes on the control and parent ion-item, but does not seem to show the red indicator line on the ion-item. Perhaps there is something I missed.

````typescript
import { ControlValueAccessor, NgControl } from '@angular/forms';
import { AfterViewInit, OnDestroy, Injector, ElementRef, NgZone, Injectable } from '@angular/core';
import { Subscription } from 'rxjs';

@Injectable()
export abstract class ValueAccessor implements ControlValueAccessor, AfterViewInit, OnDestroy {

  private statusChanges?: Subscription;

  constructor(
    protected zone: NgZone,
    protected injector: Injector,
    protected el: ElementRef
  ) { }

  onChange: (value: any) => void = () => {/**/ };
  onTouched: () => void = () => {/**/ };

  writeValue(value: any) {
    this.writeValueImpl(value);
    this.setIonicClasses(this.el);
  }

  registerOnChange(fn: (value: any) => void) {
    this.onChange = fn;
  }

  registerOnTouched(fn: () => void) {
    this.onTouched = fn;
  }

  abstract setDisabledState(isDisabled: boolean);

  abstract writeValueImpl(value: any);

  ngOnDestroy() {
    if (this.statusChanges) {
      this.statusChanges.unsubscribe();
    }
  }

  ngAfterViewInit() {
    let ngControl;
    try {
      ngControl = this.injector.get<NgControl>(NgControl); // as Type<NgControl>);
    } catch { /* No FormControl or ngModel binding */ }

    if (!ngControl) { return; }

    // Listen for changes in validity, disabled, or pending states
    if (ngControl.statusChanges) {
      this.statusChanges = ngControl.statusChanges.subscribe(() => this.setIonicClasses(this.el));
    }

    const formControl = ngControl.control;
    if (formControl) {
      const methodsToPatch = ['markAsTouched', 'markAllAsTouched', 'markAsUntouched', 'markAsDirty', 'markAsPristine'];
      methodsToPatch.forEach(method => {
        if (formControl[method]) {
          const oldFn = formControl[method].bind(formControl);
          formControl[method] = (...params) => {
            oldFn(...params);
            this.setIonicClasses(this.el);
          };
        }
      });
    }
  }

  setIonicClasses(element: ElementRef) {
    setTimeout(() => {
      const input = element.nativeElement as HTMLElement;
      const classes = this.getClasses(input);
      this.setClasses(input, classes);

      const item = input.closest('ion-item');
      if (item) {
        this.setClasses(item, classes);
      }
    }, 0);
  }

  private getClasses(element: HTMLElement) {
    const classList = element.classList;
    const classes = [];
    for (let i = 0; i < classList.length; i++) {
      const item = classList.item(i);
      if (item !== null && this.startsWith(item, 'ng-')) {
        classes.push(`ion-${item.substr(3)}`);
      }
    }
    return classes;
  }

  private setClasses(element: HTMLElement, classes: string[]) {
    const classList = element.classList;
    [
      'ion-valid',
      'ion-invalid',
      'ion-touched',
      'ion-untouched',
      'ion-dirty',
      'ion-pristine'
    ].forEach(c => classList.remove(c));

    classes.forEach(c => classList.add(c));
  }

  private startsWith(input: string, search: string): boolean {
    return input.substr(0, search.length) === search;
  }

}
```

This is primarily a clone of Ionic's `ValueAccessor` with two abstract methods `setDisabledState()` and `setValueImpl()` that need to be implemented in the custom form control class.

**Comment by liamdebeasi** [comment]
Thanks for the feature request. Are you able to provide a GitHub repo of the issues you are currently running into? This will help us evaluate the feature request.

**Comment by gmaughan** [comment]
Repository demonstrating the issues I've had is here: https://github.com/gmaughan/ionicCustomFormControls

**Comment by liamdebeasi** [comment]
Thanks! For the "Display issue" you can set `lines="full"` to get the bottom border to appear full width. For the "Required" issue, the problem is that the `ion-item` containing the file input does not have `item-interactive` which is required for the red/green status lines to show up.

These ValueAccessors are not really intended to be overridden by the user and might be going away in the future in favor of some bindings Stencil provides. That being said you should be able to manually mark fields are touched, dirty, etc and the red/green status lines should update automatically.

Does this resolve your issue?

**Comment by gmaughan** [comment]
Changing the `ion-item` that contains my custom input control to:

```html
<ion-item lines="none" class="item-interactive">
```

fixes both the formatting of the border and display of the red/green status. Thanks!

How does this attribute and class get set on an `ion-item` when it contains a standard Ionic input control? There must be some magic happening there.

I agree that the ValueAccessor is not really intended to be overridden, which is why I asked for a way to do this as a supported feature. Implementing custom controls like this is a common use case and it would be very nice to be able to get the same behaviour as standard Ionic input controls when displayed in an `ion-list`.

Perhaps an element directive would be a better solution that would allow me to attach the `setIonicClasses()`, `item-interactive` and `lines="full"` behaviour. I could then add this to my custom `app-file-input` control like so:

```html
<ion-list>
    <ion-item>
        <ion-label position="fixed">Field 2</ion-label>
        <app-file-input ionic-item-interactive [formControl]="field2" name="field2" type="text"></app-file-input>
    </ion-item>
</ion-list>
```

Either way it would be great not to have to create my own version of Ionic's ValueAccessor. I know this is probably going to break in the future.

Also note that I have to include `position="fixed"` on my `ion-label` for the label to be displayed. This isn't needed when an `ion-input` is used.

**Comment by liamdebeasi** [comment]
The way this item-interactive class gets applied is components can emit an internal `ionStyle` event. Your `app-file-input` component could do the same here. In the `ngOnInit` for your `app-file-input` component, you can do the following:

```javascript
const event = new CustomEvent('ionStyle', {
  bubbles: true,
  detail: {
    'interactive': true,
    'input': true,
   }
});
this.el.nativeElement.dispatchEvent(event);
```

There's other configuration you can do too: https://github.com/ionic-team/ionic-framework/blob/master/core/src/components/input/input.tsx#L277-L286

Does this work for your use case?

**Comment by gmaughan** [comment]
Yes, this removes the need to add attributes manually. I'm still not comfortable implementing my own version of ValueAccessor as I suspect this is likely not to work over time.

I've updated my sample repository to include the working options on Tab2 and Tab3.  https://github.com/gmaughan/ionicCustomFormControls

Thanks.
